src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -12227,7 +12230,42 @@ bool Assembler::reachable(AddressLiteral adr) {
   } else {
     disp += fudge;
   }
-  return is_simm32(disp);
+  return Assembler::is_simm32(disp);
+}
+
+bool Assembler::reachable(AddressLiteral adr) {
+  bool is_reachable = is_reachable_from(pc(), adr.target(), adr.reloc());
+  if (!is_reachable) {
+    assert(!always_reachable(adr), "sanity");
+  }
+  return is_reachable;
+}
+
+bool Assembler::always_reachable(AddressLiteral adr) {
```

Comment:
So, there is a similar block in `Assembler::is_reachable_from` -- does it make sense to reuse it then? They also are subtly different on the use of `ForceUnreachable`, would this method return different result compared to `Assembler::is_reachable_from`, and thus fail the assert in `Assembler::reachable`?

Reply:
As it is now, `ForceUnreachable` works fine.
 
There's one implicit assumption which makes it work though: temporary code buffer has to reside in code cache which is always the case now. Otherwise, `always_reachable` would have to take `pc` into account as well.

I propose to make it a requirement for the code buffer to be placed in the code cache and simplify the reachability checks based on that.